### ISSUE TITLE
Rendering: Rainbow/gradient effect results in striping when hardcoded with FFmpeg

### ISSUE THREAD

**Issue by ToshY** [libass/libass#816]
### Screenshots

Original MKV; Aegisub screenshot

![image](https://github.com/user-attachments/assets/c046addb-0891-4f3b-a96e-bdbf644cf711)

FFmpeg hardcoded (with libass 0.17.2/0.17.3); Aegisub screenshot

![image](https://github.com/user-attachments/assets/ccbbaf69-a106-43e8-80ad-54a21795d7ab)

Close-up

![image](https://github.com/user-attachments/assets/7c2f0975-383d-4a3a-a9c8-d4f9089ebf50)

### Description of the issue

Noticable striping at styles that use rainbow effects when embedding subtitles using FFmpeg.

Tried MPC-HC / Aegisub and do not see these side-effects when inspecting the source file (MKV), only after encoding and the subtitles are burned in it is noticable in the output file (MP4).

---

Source/output files [here](https://drive.google.com/drive/folders/1EKMCt4LB0s3jWLogM-_b0T1HL_2EPSHg?usp=sharing).

### libass version

0.17.3

### Is it a regression?

unknown

### ASS Sample

See source/output files [here](https://drive.google.com/drive/folders/1EKMCt4LB0s3jWLogM-_b0T1HL_2EPSHg?usp=sharing).

### Special Fonts

I uploaded or linked to the required font

### System Information

Encoded:

* Ubuntu 22.04.2 (WSL2)
* Ubuntu 22.10

Screenshots
* Win 10 - Aegisub

### Additional info

FFmpeg command; [FFMpeg `7.0.2` / libass `0.17.3` ](https://github.com/wader/static-ffmpeg/releases/tag/7.0.2)

```shell
ffmpeg -y -i input.mkv -map 0:1 -filter_complex [0:0]subtitles='input.mkv':si=0 -c:v libx264 -pix_fmt yuv420p -crf 18 -preset slow -profile:v high -level:v 4.0 -c:a copy -movflags faststart output.mp4
```

**Comment by TheOneric** [comment]
Thanks for the detailed report!

I can reproduce this with ffmpeg, but not in mpv *(including its encoder mode)* or Aegisub using libass.
Looking at the comparison screenshots, notice how apart from the small stripes inbetween segments colour of the whole subtitle lines shifts a bit. ffmpeg does not handle ASS colours completely correct. It is known to miss VSFilter-style colour mangling as documented in [`ass_types.h`](https://github.com/libass/libass/blob/b214307994be88943f609408d3182dbc9a3371ae/libass/ass_types.h#L157-L248). I also observed some issues likely related to full vs limited range independent of VSF mangling before but unfortunately wasn’t able to determine the root cause at the time and forgot about it.

The colour of the final e from “dake de” in the good screenshot is `#ee2789` and a reproduced result yields `#ec298c`; the minor difference likely stems from different encoding settings and background video. However, the bad screenshot has a significantly different colour of `#ff408b`.

I suspect this shift in colours merely makes already existing imperfection in clipping visible, while it just blends in with correct colour handling and needs to be fixed in ffmpeg. As a workaround until such a fix is available you could encode the video with [mpv’s encoder mode](https://mpv.io/manual/master/#encoding).

**Comment by astiob** [comment]
Personally, my intuitive suspect is chroma subsampling. Does this happen on non-subsampled video?

**Comment by cubicibo** [comment]
Indeed, forcing yuv444p fixes the issue. Albeit, there is the usual FFmpeg shenanigan: `-pix_fmt` is not enough, you must force the format in the filter chain, before the subtitles are burned in.

```
#> ffmpeg -i input.mkv -vf "format=yuv444p,subtitles='input.mkv':si=0" -c:v libx264 -crf 21 -preset slow -profile:v high -pix_fmt yuv420p -c:a copy -movflags faststart output.mp4
```

**Comment by cubicibo** [comment]
Supposedly you can tag the pixel format per stream (input, output) with `-pix_fmt` but it does not really work/is poorly documented. You should opt to use the filter and to blend in 444 whenever you are dealing with colourful subtitles.

libass outputs 32-bit RGBA overlays that can be losslessly converted to YUV444 + Alpha. The FFmpeg burn-in filter truncate the overlay chroma to match the chroma planes of the input video. HandBrake has the same issue despite not using FFmpeg burn-in filters. The problem is quite spread.

**Comment by cubicibo** [comment]
HandBrake blends every `ASS_Image` one after the other, converting them on the fly to the video planar format. The chroma sample is taken as-is, and neighbouring values are just disregarded.
![mpv-shot0001](https://github.com/user-attachments/assets/0b852077-8485-4dba-9234-abc020c4cbd9)

FFmpeg evidently takes more care, and fails more subtlety. The chroma sample seems to be selected more wisely, hence the thinner vertical stripes.

**Comment by TheOneric** [comment]
> I also observed some issues likely related to full vs limited range independent of VSF mangling before but unfortunately wasn’t able to determine the root cause at the time and forgot about it.

Turns out this too is fixed by forcing yuv444p and thus presumably the same issue. Just because the affected types consisted of many thin vector drawings the resulting effect appeared like a general colour shift rather than striping.

**Comment by TheOneric** [comment]
I reported this to ffmpeg: https://trac.ffmpeg.org/ticket/11140#no1

Btw it’s probably a good idea to use ffmpeg’s `ass` filter instead of `subtitles` to hardsub ASS; judging by respective issues still being open it appears `subtitles` still forces usage of the simple shaper which will break certain scripts: https://trac.ffmpeg.org/ticket/8738 and https://trac.ffmpeg.org/ticket/10384

HandBrake (understandably) wants logs and detailed version info for their issues; can someone using HandBrake report it to them too?

**Comment by cubicibo** [comment]
I'll handle the HandBrake PR. I don't have the will to send a patch to FFmpeg. Ideally, `vf_subtitles` and `vf_ass` should use `vf_overlay` internal functions as they are the most efficient and versatile. However, a major refactoring will doom the patch from being merged in anyone's lifetime.

**Comment by MasterInQuestion** [comment]
Per my observation, alike colorspace conversions shouldn't much impair quality.
    When properly utilized.

    See also: https://trac.ffmpeg.org/wiki/colorspace#Examples

**Comment by arch1t3cht** [comment]
> Btw it’s probably a good idea to use ffmpeg’s ass filter instead of subtitles to hardsub ASS; judging by respective issues still being open it appears subtitles still forces usage of the simple shaper which will break certain scripts: https://trac.ffmpeg.org/ticket/8738 and https://trac.ffmpeg.org/ticket/10384

I was looking through this issue while researching some things and in the process also found out that this was fixed in ffmpeg now: <https://code.ffmpeg.org/FFmpeg/FFmpeg/commit/b08c9c5f533e7dced089d12d4526301cc200c16e> (part of ffmpeg 9.0).

As far as I can tell this means that `vf=subtitles` is now strictly superior to `vf=ass` in a normal command-line invocation, at least assuming no bugs in ffmpeg's ass file parser.

**Comment by ToshY** [comment]
> > Btw it’s probably a good idea to use ffmpeg’s ass filter instead of subtitles to hardsub ASS; judging by respective issues still being open it appears subtitles still forces usage of the simple shaper which will break certain scripts: https://trac.ffmpeg.org/ticket/8738 and https://trac.ffmpeg.org/ticket/10384
> 
> I was looking through this issue while researching some things and in the process also found out that this was fixed in ffmpeg now: https://code.ffmpeg.org/FFmpeg/FFmpeg/commit/b08c9c5f533e7dced089d12d4526301cc200c16e (part of ffmpeg 9.0).

I got a notification on this thread and thought the isue might be fixed in FFmpeg 9.0 seeing your reply, but the shaping `simple`/`complex` is not directly related to this 🙁 

(tested the source files with FFmpeg `8.1.2` and `9.0` again, still having same striping as was in `7.0.2`; I also now see there's no apparent activity in https://trac.ffmpeg.org/ticket/11140).

**Comment by arch1t3cht** [comment]
Yes, sorry, I should have clarified this. The striping issue is not fixed in FFmpeg, but the shaping issue Oneric mentioned is.
